### PR TITLE
Consolidate RobotState benchmarks in single file

### PR DESCRIPTION
**Pull request by marioprats**
### Description

This change consolidates `robot_state_benchmark.cpp` and `robot_state_jacobian_benchmark.cpp` into a single `robot_state_benchmark.cpp` file, and adds further benchmarks for RobotState construction, copy, and forward kinematics. This is in preparation for a follow-up change to update memory management in `RobotState`.
The matrix multiplication / inverse tests have been fully rewritten into a gmock BENCHMARK for consistency.

Output of running:
```
./build/moveit_core/robot_state/robot_state_benchmark 2>/dev/null
---------------------------------------------------------------------------------
Benchmark                                       Time             CPU   Iterations
---------------------------------------------------------------------------------
MultiplyAffineTimesMatrix/10000000           55.7 ms         55.7 ms           13
MultiplyMatrixTimesMatrix/10000000           57.3 ms         57.3 ms           12
MultiplyIsometryTimesIsometry/10000000       57.5 ms         57.5 ms           12
InverseIsometry3d/10000000                    107 ms          107 ms            7
InverseAffineIsometry/10000000               23.3 ms         23.3 ms           30
InverseAffine/10000000                       52.7 ms         52.7 ms           14
InverseMatrix4d/10000000                      109 ms          109 ms            6
RobotStateConstruct/100                     0.005 ms        0.005 ms       123429
RobotStateConstruct/1000                    0.054 ms        0.054 ms        12827
RobotStateConstruct/10000                   0.553 ms        0.553 ms         1250
RobotStateCopy/100                          0.006 ms        0.006 ms       114604
RobotStateCopy/1000                         0.061 ms        0.061 ms        11284
RobotStateCopy/10000                        0.615 ms        0.615 ms         1141
RobotStateUpdate/10                         0.026 ms        0.026 ms        26048
RobotStateUpdate/100                        0.269 ms        0.269 ms         2615
RobotStateUpdate/1000                        2.66 ms         2.66 ms          262
RobotStateForwardKinematics/10              0.023 ms        0.023 ms        29936
RobotStateForwardKinematics/100             0.234 ms        0.234 ms         2996
RobotStateForwardKinematics/1000             2.34 ms         2.34 ms          304
MoveItJacobian                                258 ns          258 ns      2714295
KdlJacobian                                  1106 ns         1107 ns       630134

```

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 1]
```diff
@@ -32,132 +32,338 @@
  *  POSSIBILITY OF SUCH DAMAGE.
  *********************************************************************/
 
-/* Author: Robert Haschke */
+/* Author: Mario Prats, Robert Haschke */
```
Comment (sea-bass) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:35]: Maybe put yourself 2nd to make clearer who the original author is.

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 2]
```diff
@@ -32,132 +32,338 @@
...
+#include <random_numbers/random_numbers.h>
 
-// Helper class to measure time within a scoped block and output the result
-class ScopedTimer
-{
-  const char* const msg_;
-  double* const gold_standard_;
-  const std::chrono::time_point<std::chrono::steady_clock> start_;
+// Robot and planning group for benchmarks, except for RobotStateUpdate and RobotStateForwardKinematics, which will use
+// the pr2.
```
Comment (sea-bass) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:48]: Ultra nitpick -- also, should these be called `DEFAULT_TEST_ROBOT` etc. instead?

```suggestion
// the PR2.
```

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 3]
```diff
@@ -32,132 +32,338 @@
...
+  {
+    st.PauseTiming();
+    kinematic_state.setToRandomPositions(jmg, rng);
+    kinematic_state.updateLinkTransforms();
+    KDL::Jacobian jacobian(kdl_chain.getNrOfJoints());
+    KDL::JntArray kdl_q;
+    kdl_q.resize(kdl_chain.getNrOfJoints());
+    kinematic_state.copyJointGroupPositions(jmg, &kdl_q.data[0]);
+    st.ResumeTiming();
+    jacobian_solver.JntToJac(kdl_q, jacobian);
```
Comment (sea-bass) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:350]: worth adding the comment here as well that we're timing only the Jacobian?

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 4]
```diff
@@ -32,132 +32,338 @@
...
+BENCHMARK(InverseIsometry3d)->Arg(1e7)->Unit(benchmark::kMillisecond);
+BENCHMARK(InverseAffineIsometry)->Arg(1e7)->Unit(benchmark::kMillisecond);
+BENCHMARK(InverseAffine)->Arg(1e7)->Unit(benchmark::kMillisecond);
+BENCHMARK(InverseMatrix4d)->Arg(1e7)->Unit(benchmark::kMillisecond);
+
+BENCHMARK(RobotStateConstruct)->RangeMultiplier(10)->Range(100, 10000)->Unit(benchmark::kMillisecond);
+BENCHMARK(RobotStateCopy)->RangeMultiplier(10)->Range(100, 10000)->Unit(benchmark::kMillisecond);
+BENCHMARK(RobotStateUpdate)->RangeMultiplier(10)->Range(10, 1000)->Unit(benchmark::kMillisecond);
+BENCHMARK(RobotStateForwardKinematics)->RangeMultiplier(10)->Range(10, 1000)->Unit(benchmark::kMillisecond);
+
```
Comment (sea-bass) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:367]: I feel like a lot of these could be made into `constexpr`s to clarify intent and reduce copypasta errors if/when changes must be made

File: moveit_core/robot_state/test/robot_state_benchmark.cpp (hunk 1 above)
Comment (marioprats) [reply moveit_core/robot_state/test/robot_state_benchmark.cpp]: sg

File: moveit_core/robot_state/test/robot_state_benchmark.cpp (hunk 2 above)
Comment (marioprats) [reply moveit_core/robot_state/test/robot_state_benchmark.cpp]: I changed these to:
```
constexpr char PANDA_TEST_ROBOT[] = "panda";
constexpr char PANDA_TEST_GROUP[] = "panda_arm";
constexpr char PR2_TEST_ROBOT[] = "pr2";
constexpr char PR2_TIP_LINK[] = "r_wrist_roll_link";
```

**Review by sea-bass — APPROVED** [review]

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 5]
```diff
@@ -32,132 +32,342 @@
  *  POSSIBILITY OF SUCH DAMAGE.
  *********************************************************************/
 
-/* Author: Robert Haschke */
+/* Author: Robert Haschke, Mario Prats */
+
+// To run this benchmark, 'cd' to the build/moveit_core/robot_state directory and directly run the binary.
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:37]: Consider clarifying what is benchmarked here

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 6]
```diff
@@ -32,132 +32,342 @@
...
-  ScopedTimer(const char* msg = "", double* gold_standard = nullptr)
-    : msg_(msg), gold_standard_(gold_standard), start_(std::chrono::steady_clock::now())
+// Number of iterations to use in matrix multiplication / inversion benchmarks.
+constexpr int MATRIX_OPS_N_ITERATIONS = 1e7;
+
+static void multiplyAffineTimesMatrix(benchmark::State& st)
+{
+  int n_iters = st.range(0);
+  Eigen::Isometry3d isometry = Eigen::Translation3d(1, 2, 3) *
+                               Eigen::AngleAxisd(0.13 * M_PI, Eigen::Vector3d::UnitX()) *
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:60]: Can you add an explanation for the magic numbers 0.13, 0.29 and 0.42?

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 7]
```diff
@@ -32,132 +32,342 @@
...
+constexpr char PR2_TIP_LINK[] = "r_wrist_roll_link";
 
-public:
-  // if gold_standard is provided, a relative increase/decrease is shown too
-  ScopedTimer(const char* msg = "", double* gold_standard = nullptr)
-    : msg_(msg), gold_standard_(gold_standard), start_(std::chrono::steady_clock::now())
+// Number of iterations to use in matrix multiplication / inversion benchmarks.
+constexpr int MATRIX_OPS_N_ITERATIONS = 1e7;
+
+static void multiplyAffineTimesMatrix(benchmark::State& st)
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:56]: Add brief description

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 8]
```diff
@@ -32,132 +32,342 @@
...
+    {
+      Eigen::Affine3d result;
+      benchmark::DoNotOptimize(result = isometry.affine() * isometry.matrix());
+      benchmark::ClobberMemory();
+    }
   }
+}
 
-  ~ScopedTimer()
+static void multiplyMatrixTimesMatrix(benchmark::State& st)
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:74]: Add brief description?

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 9]
```diff
@@ -32,132 +32,342 @@
...
+    {
+      Eigen::Matrix4d result;
+      benchmark::DoNotOptimize(result = isometry.matrix() * isometry.matrix());
+      benchmark::ClobberMemory();
+    }
+  }
+}
 
-    if (gold_standard_)
+static void multiplyIsometryTimesIsometry(benchmark::State& st)
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:92]: Add brief description?

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 10]
```diff
@@ -32,132 +32,342 @@
...
+      benchmark::DoNotOptimize(result = isometry * isometry);
+      benchmark::ClobberMemory();
     }
-    std::cerr << '\n';
   }
-};
+}
 
-class Timing : public testing::Test
+static void inverseIsometry3d(benchmark::State& st)
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:110]: Add brief description?

File: moveit_core/robot_state/test/robot_state_benchmark.cpp [hunk 11]
```diff
@@ -32,132 +32,342 @@
...
+    for (int i = 0; i < n_iters; ++i)
+    {
+      Eigen::Isometry3d result;
+      benchmark::DoNotOptimize(result = isometry.inverse());
+      benchmark::ClobberMemory();
+    }
   }
+}
+
+static void inverseAffineIsometry(benchmark::State& st)
```
Comment (sjahr) [inline moveit_core/robot_state/test/robot_state_benchmark.cpp:128]: Add brief description?

**Review by sjahr — COMMENTED** [review]
Thanks for this cleanup! The unification of the files makes sense to me. Could you add documentation for this file to make maintainability easier in the future? I think ~ one brief comment for every function + benchmark and explanation for some parameters is sufficient.

File: moveit_core/robot_state/test/robot_state_benchmark.cpp (hunk 5 above)
Comment (marioprats) [reply moveit_core/robot_state/test/robot_state_benchmark.cpp:38]: Added a general comment here, and additional comments before each of the benchmarks.

File: moveit_core/robot_state/test/robot_state_benchmark.cpp (hunk 6 above)
Comment (marioprats) [reply moveit_core/robot_state/test/robot_state_benchmark.cpp]: Moved this into a helper function and added a comment saying these are arbitrary test values.

**Review by sjahr — APPROVED** [review]
Thanks!

**Pull request merged**